index/scorch/snapshot_segment.go

```diff
@@ -66,6 +66,53 @@ func (s *SegmentSnapshot) LiveSize() int64 {
 	return int64(s.Count())
 }
 
+func (s *SegmentSnapshot) HasVector() bool {
+	if fieldVecs, ok := s.stats.Fetch()["num_vecs"]; ok {
+		return len(fieldVecs) > 0
+	}
+
+	return false
+}
+
+func (s *SegmentSnapshot) FullVectorsByteSize() uint64 {
+	vecByteSize, ok := s.stats.Fetch()["vec_byte_size"]
+	if !ok {
+		return 0
+	}
+
+	var rv uint64
+	for _, size := range vecByteSize {
+		rv += size
+	}
+
+	return rv
+}
+
+func (s *SegmentSnapshot) LiveVectorsByteSize() uint64 {
+	fullSize := s.FullVectorsByteSize()
+
+	var deletedDocsCount uint64
+	if s.deleted != nil {
+		deletedDocsCount = s.deleted.GetCardinality()
+	}
+
+	if deletedDocsCount == 0 {
+		return fullSize
+	}
+
+	// # Live Vector Size Estimation
```

Comment:
i have a doubt here, from what i understand over here the assumption is that each deleted document has a vector and the vector size is distributed equally (equal distribution of vectors over all docs in the segment). 

in a segment having 3 fields 1 text and 2 vectors - if the workload pattern (upserts) was such that the deleted documents in this segment corresponding to only the text field content and the two vector field contents' documents weren't invalidated (because the updates of vectors is not a frequent thing lets say), then we could perhaps end up making incorrect decision over here. 

instead, maybe what can be done is 
1. track the min and max docIDs across all the vector fields.
2. compare this range with the deleted bitmap's range (which also has min and max APIs).
3. make the decision of how much is the live vector size of this segment based on the overlap of these ranges. for eg, if there is no overlap in the two ranges then the live vector size of the segment wouldn't be affected.

Reply:
>in a segment having 3 fields 1 text and 2 vectors - if the workload pattern (upserts) was such that the deleted documents in this segment corresponding to only the text field content and the two vector field contents' documents weren't invalidated (because the updates of vectors is not a frequent thing lets say), then we could perhaps end up making incorrect decision over here.

@Thejas-bhat raises a good point, am I misremembering in thinking - per our last conversation we decide not to do the live size but just use the actual size in making decisions.

Reply:
@abhinavdangeti  In our last conversation, we decided to not use liveSize , because the solution that I proposed needed scorch to know dimensions of vectors for each field. 

But, with this new approach, where we don't need to know dimensions. 
I think we should use LiveSize as an eligibility criteria for segments to be merged

Reply:
@Thejas-bhat  @abhinavdangeti  

> then we could perhaps end up making incorrect decision over here.
Can you elaborate on this line 

My notion of a mutation is that , even if editing is done only in text field, we mark entire content of the doc in old segments as invalid, am I right here ? 

As part of a mutation, KV will send us the new doc ( same ID,  unedited + edited content both ) , right ?

Reply:
You're right Shaad - but I think, @Thejas-bhat 's point is that these docs that were updated or deleted never had any vector content in them to begin with.